server/js/src/main/scala/org/http4s/server/middleware/CSRFPlatform.scala

```diff
@@ -0,0 +1,158 @@
+    * if constructed with a bad key.
+    */
+  def signToken[M[_]](rawToken: String)(implicit F: Async[M]): M[CSRFToken] = {
+    val joined = rawToken + "-" + clock.millis().toString()
+    signTokenNonce(joined).map(ba => lift(joined + "-" + encodeHexString(ba)))
+  }
+
+  /** Decode our CSRF token, check the signature
+    * and extract the original token string to sign
+    */
+  def extractRaw[M[_]](rawToken: String)(implicit F: Async[M]): M[Either[CSRFCheckFailed, String]] =
+    rawToken.split("-") match {
+      case Array(raw, nonce, signed) =>
+        signTokenNonce(raw + "-" + nonce).map { out =>
+          decodeHexString(signed) match {
+            case Some(decoded) =>
+              if (digestIsEqual(out, decoded))
+                Right(raw)
+              else
+                Left(CSRFCheckFailed)
+            case None =>
+              Left(CSRFCheckFailed)
+          }
+        }
+      case _ =>
+        F.pure(Left(CSRFCheckFailed))
+    }
+
+}
+
+private[middleware] trait CSRFSingletonPlatform { self: CSRF.type =>
+  type SecretKey = org.http4s.js.crypto.CryptoKey
+  val SigningAlgo: KeyAlgorithm =
+    js.Dynamic.literal(name = "HMAC", hash = "SHA-1").asInstanceOf[KeyAlgorithm]
+
+  /** A Constant-time string equality */
+  def isEqual(s1: String, s2: String): Boolean =
+    digestIsEqual(s1.getBytes(StandardCharsets.UTF_8), s2.getBytes(StandardCharsets.UTF_8))
+
+  /** Generate an unsigned CSRF token from a `SecureRandom` */
+  private[middleware] def genTokenString: String = {
+    val arrayBuffer = new Int8Array(CSRFTokenLength)
+    webcrypto.getRandomValues(arrayBuffer.asInstanceOf[ArrayBufferView])
+    val bytes = new Array[Byte](CSRFTokenLength)
+    TypedArrayBuffer.wrap(arrayBuffer).get(bytes)
+    encodeHexString(bytes)
+  }
+
+  /** Generate a signing Key for the CSRF token */
+  def generateSigningKey[F[_]]()(implicit F: Async[F]): F[SecretKey] =
+    F.fromPromise {
+      F.delay {
+        subtle.generateKey(SigningAlgo, false, js.Array(KeyUsage.sign))
+      }
+    }.map(_.asInstanceOf[SecretKey])
+
+  /** Build a new HMACSHA1 Key for our CSRF Middleware
+    * from key bytes. This operation is unsafe, in that
+    * any amount less than 20 bytes will throw an exception when loaded
+    * into `Mac`. Any keys larger than 64 bytes are just hashed.
+    *
+    * For more information, refer to: https://tools.ietf.org/html/rfc2104#section-3
+    *
+    * Use for loading a key from a config file, after having generated
+    * one safely
+    */
+  def buildSigningKey[F[_]](array: Array[Byte])(implicit F: Async[F]): F[SecretKey] =
+    F.fromPromise {
+      F.delay {
+        subtle.importKey(
+          KeyFormat.raw,
+          array.toTypedArray.asInstanceOf[BufferSource],
+          SigningAlgo,
+          false,
+          js.Array(KeyUsage.sign)
+        )
+      }
+    }.map(_.asInstanceOf[SecretKey])
+
+  private[middleware] def digestIsEqual(digesta: Array[Byte], digestb: Array[Byte]) =
```

Comment:
What is platform specific about this?

Reply:
Because JVM just has to do `MessageDigest.isEqual(out, decoded)`